### PR TITLE
[SPARK-23698][Python] Resolve undefined names in Python 3

### PR DESCRIPTION
**Pull request by cclauss**
## What changes were proposed in this pull request?

Fix issues arising from the fact that builtins __file__, __long__, __raw_input()__, __unicode__, __xrange()__, etc. were all removed from Python 3.  __Undefined names__ have the potential to raise [NameError](https://docs.python.org/3/library/exceptions.html#NameError) at runtime.

## How was this patch tested?
* $ __python2 -m flake8 . --count --select=E9,F82 --show-source --statistics__
* $ __python3 -m flake8 . --count --select=E9,F82 --show-source --statistics__

@holdenk

flake8 testing of https://github.com/apache/spark on Python 3.6.3

$ __python3 -m flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__
```
./dev/merge_spark_pr.py:98:14: F821 undefined name 'raw_input'
    result = raw_input("\n%s (y/n): " % prompt)
             ^
./dev/merge_spark_pr.py:136:22: F821 undefined name 'raw_input'
    primary_author = raw_input(
                     ^
./dev/merge_spark_pr.py:186:16: F821 undefined name 'raw_input'
    pick_ref = raw_input("Enter a branch name [%s]: " % default_branch)
               ^
./dev/merge_spark_pr.py:233:15: F821 undefined name 'raw_input'
    jira_id = raw_input("Enter a JIRA id [%s]: " % default_jira_id)
              ^
./dev/merge_spark_pr.py:278:20: F821 undefined name 'raw_input'
    fix_versions = raw_input("Enter comma-separated fix version(s) [%s]: " % default_fix_versions)
                   ^
./dev/merge_spark_pr.py:317:28: F821 undefined name 'raw_input'
            raw_assignee = raw_input(
                           ^
./dev/merge_spark_pr.py:430:14: F821 undefined name 'raw_input'
    pr_num = raw_input("Which pull request would you like to merge? (e.g. 34): ")
             ^
./dev/merge_spark_pr.py:442:18: F821 undefined name 'raw_input'
        result = raw_input("Would you like to use the modified title? (y/n): ")
                 ^
./dev/merge_spark_pr.py:493:11: F821 undefined name 'raw_input'
    while raw_input("\n%s (y/n): " % pick_prompt).lower() == "y":
          ^
./dev/create-release/releaseutils.py:58:16: F821 undefined name 'raw_input'
    response = raw_input("%s [y/n]: " % msg)
               ^
./dev/create-release/releaseutils.py:152:38: F821 undefined name 'unicode'
        author = unidecode.unidecode(unicode(author, "UTF-8")).strip()
                                     ^
./python/setup.py:37:11: F821 undefined name '__version__'
VERSION = __version__
          ^
./python/pyspark/cloudpickle.py:275:18: F821 undefined name 'buffer'
        dispatch[buffer] = save_buffer
                 ^
./python/pyspark/cloudpickle.py:807:18: F821 undefined name 'file'
        dispatch[file] = save_file
                 ^
./python/pyspark/sql/conf.py:61:61: F821 undefined name 'unicode'
        if not isinstance(obj, str) and not isinstance(obj, unicode):
                                                            ^
./python/pyspark/sql/streaming.py:25:21: F821 undefined name 'long'
    intlike = (int, long)
                    ^
./python/pyspark/streaming/dstream.py:405:35: F821 undefined name 'long'
        return self._sc._jvm.Time(long(timestamp * 1000))
                                  ^
./sql/hive/src/test/resources/data/scripts/dumpdata_script.py:21:10: F821 undefined name 'xrange'
for i in xrange(50):
         ^
./sql/hive/src/test/resources/data/scripts/dumpdata_script.py:22:14: F821 undefined name 'xrange'
    for j in xrange(5):
             ^
./sql/hive/src/test/resources/data/scripts/dumpdata_script.py:23:18: F821 undefined name 'xrange'
        for k in xrange(20022):
                 ^
20    F821 undefined name 'raw_input'
20
```

### CODE REVIEW

File: dev/create-release/releaseutils.py
Comment (HyukjinKwon) [reply dev/create-release/releaseutils.py]: and shall we do this with if-else of Python version to be consistent with other places?

Comment (cclauss) [reply dev/create-release/releaseutils.py]: “Where practical, apply the Python porting best practice: [Use feature detection instead of version detection](https://docs.python.org/3/howto/pyporting.html#use-feature-detection-instead-of-version-detection).”

Comment (HyukjinKwon) [reply dev/create-release/releaseutils.py]: Shall we keep this consistent for now? There are a hell of a lot things to fix in PySpark in that way.

File: dev/create-release/releaseutils.py
Comment (HyukjinKwon) [inline dev/create-release/releaseutils.py]: Two spaces before the inlined comment ..

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by cclauss** [comment]
@HyukjinKwon Was there something more to do on this one?

**Review by HyukjinKwon — COMMENTED** [review]
shall we wait for `https://github.com/cloudpipe/cloudpickle/pull/163`? I want to see more feedback.

**Review by felixcheung — COMMENTED** [review]
that cloudpickle PR seems independent on the changes here though?

File: python/pyspark/cloudpickle.py
Comment (HyukjinKwon) [inline python/pyspark/cloudpickle.py]: I think this one is actually related with cloudpickle's PR.

I was trying to (exactly) match this file to a specific version of cloudpickle (which is currently 0.4.3 - https://github.com/cloudpipe/cloudpickle/releases/tag/v0.4.3). So, I thought we could wait for more feedback there. 

At least, I was thinking that we should match it to https://github.com/cloudpipe/cloudpickle/tree/0.4.x If that one is merged, I could backport that change into 0.4.x branch.

File: dev/create-release/releaseutils.py
Comment (BryanCutler) [inline dev/create-release/releaseutils.py]: could you just put this above:
```
if sys.version > '3':
    unicode = str
```
and then just do?
```
author = unidecode.unidecode(unicode(author)).strip()
```

I don't think you need to specify "UTF-8" because either way it will be a unicode object

**Review by BryanCutler — COMMENTED** [review]
Thanks @cclauss for the PR, +1 for syncing the cloudpickle change.

**Comment by cclauss** [comment]
The cloudpickle change is now merged upstream.  I would like to avoid @BryanCutler suggestion above because Unicode is really touchy in Python 2 and a lot can change based on [sys.setdefaultencoding()](https://docs.python.org/2/library/sys.html#sys.setdefaultencoding) so I would like to avoid assumptions and make as small modifications as possible.

**Comment by cclauss** [comment]
Other issues / suggestions?

File: python/pyspark/cloudpickle.py
Comment (cclauss) [reply python/pyspark/cloudpickle.py]: This was merged upstream.

**Comment by HyukjinKwon** [comment]
im preparing to backport and release cloudpuckle. please give me some more days.

**Comment by HyukjinKwon** [comment]
ok to tesr

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by felixcheung** [comment]
any update?

File: python/pyspark/streaming/dstream.py
Comment (HyukjinKwon) [inline python/pyspark/streaming/dstream.py]: Can you add a test for it? Seems only used once and shouldn't be difficult to add a test.

File: python/pyspark/sql/conf.py
Comment (HyukjinKwon) [inline python/pyspark/sql/conf.py]: This is fine since we rely on short-circuiting.

File: dev/merge_spark_pr.py
Comment (HyukjinKwon) [inline dev/merge_spark_pr.py]: Does this script run with Python 3+ now?

File: dev/create-release/releaseutils.py
Comment (HyukjinKwon) [inline dev/create-release/releaseutils.py]: Does this script work in Python 3+ now?

File: dev/merge_spark_pr.py
Comment (cclauss) [reply dev/merge_spark_pr.py]: It creates a new function called __raw_input()__ that is identical to the builtin __input()__.

File: dev/create-release/releaseutils.py
Comment (cclauss) [reply dev/create-release/releaseutils.py]: It creates a new function called __raw_input()__ that is identical to the builtin __input()__.

**Comment by HyukjinKwon** [comment]
@cclauss, mind updating this PR?

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by BryanCutler** [comment]
retest this please

**Comment by BryanCutler** [comment]
I don't think it's your PR, let's try again

**Comment by BryanCutler** [comment]
retest this please

File: dev/create-release/releaseutils.py
Comment (holdenk) [reply dev/create-release/releaseutils.py]: @BryanCutler I think we do need to specify it because it won't be unicode type in Python 2, we get these from run_cmd which call's Popen communicate which returns a regular string.

File: python/pyspark/streaming/dstream.py
Comment (holdenk) [reply python/pyspark/streaming/dstream.py]: Adding a call to slice in `tests.py` should be enough to test this. I'm surprised we haven't caught this before, but I suppose this isn't a very frequently exercised code path.

**Review by holdenk — COMMENTED** [review]
Thank you so much for working on this! I really appreciate it. One comment about how to do the test for `long = int` by testing the one place that uses it (slice) in DStream. A good follow up PR to this could be adding `flake8` to the `./dev/lint-python` script to we don't have similar mistakes introduced in the future.

Thanks again! :)

File: dev/create-release/releaseutils.py
Comment (cclauss) [reply dev/create-release/releaseutils.py]: Yes... My other worry in Py2 would be if [__sys.setdefaultencoding()__](https://docs.python.org/2/library/sys.html#sys.setdefaultencoding) was set to somthing other that utf-8.  That method was also thankfully dropped in Py3.

Comment (BryanCutler) [reply dev/create-release/releaseutils.py]: My thought was that we are first casting `author` this to unicode already with `unicode(author)` and it doesn't really matter if it is "UTF-8" or not because we then immediately decode it into ASCII with `unidecode`, which can handle it even it it wasn't "UTF-8", so the end result should be the same I believe.  It was just to clean up a little, so not a big deal either way.  The way it is now replicates the old behavior, so it's probably safer.

File: python/pyspark/sql/conf.py
Comment (HyukjinKwon) [reply python/pyspark/sql/conf.py]: Nope

**Comment by holdenk** [comment]
Ping @cclauss do you have some time to update this PR with the feedback? We've got a code freeze coming up "soon" and it would be nice to be able to get this in before code freeze.

File: dev/create-release/releaseutils.py
Comment (cclauss) [reply dev/create-release/releaseutils.py]: > The way it is now [...] it's probably safer.

Let's agree to leave this as is in this PR.  EOL of Python 2 in 500 daze away so safe is better.

File: python/pyspark/streaming/tests.py
Comment (cclauss) [inline python/pyspark/streaming/tests.py]: @holdenk Comments please on __test_slice()__ and on the test results 
https://github.com/apache/spark/pull/20838#issuecomment-408566860

**Review by HyukjinKwon — APPROVED** [review]

**Comment by cclauss** [comment]
@holdenk Did we miss the window?  I still count 10 undefined names in this repo.

**Comment by HyukjinKwon** [comment]
@cclauss, please go ahead and fix them together.

**Comment by cclauss** [comment]
Several are fixed it this PR.  Some others are not fixable without resorting to using __#noqa__.

**Comment by holdenk** [comment]
So this is failing Spark unit tests, not PySpark so my guess is the failures are unrelated. Jenkins retest this please.

File: python/pyspark/streaming/tests.py
Comment (holdenk) [inline python/pyspark/streaming/tests.py]: please remove this comment.

**Comment by holdenk** [comment]
It looks pretty reasonable, I do want to make sure that those test failures were spurious so lets wait for the Jenkins re-run.

**Comment by holdenk** [comment]
Ok only failure is in Kafka, very likely unrelated. Jenkins retest this please.

**Review by HyukjinKwon — APPROVED** [review]
LGTM but let me leave it to @holdenk for a final look and merge

**Comment by cclauss** [comment]
May I please break this PR into pieces as I did in #21959 and #21960 so that we can isolate why tests keep failing?  These issues were all discovered using the same (flake8 F821) technique but they are in separate files and cover different undefined names so they could rightfully be part of separate PRs.

**Comment by cclauss** [comment]
jenkins, retest this please

**Comment by HyukjinKwon** [comment]
I don't think we should split this anymore. you can just fix the test here.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by cclauss** [comment]
@HyukjinKwon Your advise on next steps?

**Comment by cclauss** [comment]
jenkins, retest this please

**Comment by holdenk** [comment]
So if the issue was a timeout last time, we can ask "Jenkins retest this please" and see if not now works. If that isn't working for you another way we can trigger Jenkins retest is to merge in master or rebase and push so Jenkins sees a new commit.

**Comment by holdenk** [comment]
Looking at Jenkins though I can see somethings which don't appear to be timeout failures, e.g.

> Traceback (most recent call last):
>   File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/streaming/tests.py", line 224, in test_slice
>     self._test_func(input, func, expected)
>   File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/streaming/tests.py", line 147, in _test_func
>     input = [self.sc.parallelize(d, 1) for d in input]
>   File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/context.py", line 494, in parallelize
>     c = list(c)    # Make it a list so we can compute its length
> TypeError: 'datetime' object is not iterable
> 

Have you run the tests locally and do they pass?

Looking at `_test_func` it appears that it expects a list of lists for input rather than list, and each list inside of the list of lists represents one time slide of the `DStream`. Hope that helps, maybe looking at some of the other tests in the same file could also be of use. If you get stuck let me know :)

**Comment by HyukjinKwon** [comment]
> @HyukjinKwon Your advise on next steps?

Shall we make the tests passed? 

> Have you run the tests locally and do they pass?

Same question.

**Comment by HyukjinKwon** [comment]
Hm, why https://github.com/apache/spark/pull/20838/commits/b178587909be038b8bc5c7254fb2a384ae61a7e4 was reverted?

**Comment by cclauss** [comment]
It was reverted because [__slice_test()__](#22128) was causing the build to fail.

**Comment by HyukjinKwon** [comment]
Shell we fix it here?

**Comment by cclauss** [comment]
This is not working at all...  I am wasting way too much time.  5+ months and 80+ comments for 12 lines of code is 

I do not have the skills to solve the following undefined name 'long' in a satisfactory manner:
```
./python/pyspark/streaming/dstream.py:405:35: F821 undefined name 'long'
        return self._sc._jvm.Time(long(timestamp * 1000))
                                  ^
```
If someone with more skills would be willing to take that one undefined name off my plate and solve it with a test in a separate PR then I would be grateful.  I will study that PR carefully and can then proceed with the others that are in this PR.

My recommended fix is at https://github.com/apache/spark/pull/20838/files#diff-6c576c52abc0624ccb6a2f45828dc6a7 and my proposed test (it is failing!) is immediately following.

**Comment by BryanCutler** [comment]
Hi @cclauss , sorry for the frustration.  I looked into the test, and it was kind of a pain to get it working right - which is probably why it wasn't done in the first place ;)

Here are my modifications for `test_slice` and it seems to pass py3 fine

```python
def test_slice(self):
        """Basic operation test for DStream.slice."""
        import datetime as dt
        self.ssc = StreamingContext(self.sc, 1.0)
        self.ssc.remember(4.0)
        input = [[1], [2], [3], [4]]
        stream = self.ssc.queueStream([self.sc.parallelize(d, 1) for d in input])

        time_vals = []

        def get_times(t, rdd):
            if rdd and len(time_vals) < len(input):
                time_vals.append(t)

        stream.foreachRDD(get_times)

        self.ssc.start()
        self.wait_for(time_vals, 4)
        begin_time = time_vals[0]

        def get_sliced(begin_delta, end_delta):
            begin = begin_time + dt.timedelta(seconds=begin_delta)
            end = begin_time + dt.timedelta(seconds=end_delta)
            rdds = stream.slice(begin, end)
            result_list = [rdd.collect() for rdd in rdds]
            return [r for result in result_list for r in result]

        self.assertEqual(set([1]), set(get_sliced(0, 0)))
        self.assertEqual(set([2, 3]), set(get_sliced(1, 2)))
        self.assertEqual(set([2, 3, 4]), set(get_sliced(1, 4)))
        self.assertEqual(set([1, 2, 3, 4]), set(get_sliced(0, 4)))
```

If you want to put that in, I have some time now and can help you get this merged or if you prefer I can finish it up and still assign to you.

```p

**Comment by cclauss** [comment]
Thanks massively for this.  I doubt that I _ever_ would have gotten to that on my own.  This is a test so my proposal would be that _you create a separate PR_ so that we are all assured that it passes in the current codebase.  Once that PR has been merged, I can come back and finish this PR.  Thanks again.

**Comment by felixcheung** [comment]
Or that Bryan opens a PR on your branch? that usually would be easier to get *this* PR through, just my 2c.

**Comment by HyukjinKwon** [comment]
+1 for ^. We can credit to multiple persons now anyway.

**Review by BryanCutler — COMMENTED** [review]
LGTM, just one minor change. Could you also update the PR title to include a [PYTHON] tag, and from the description remove the line crossed out "Where practical..." and "Please review http://spark.apache.org/contributing.html before opening a pull request."?

**Review by BryanCutler — APPROVED** [review]
LGTM. If you could just update the title and description mentioned in https://github.com/apache/spark/pull/20838#pullrequestreview-148163149 I think this is good to go

**Comment by BryanCutler** [comment]
merged to master, thanks @cclauss !

**Pull request closed without merging**
